Editor/EditorCore/ProBuilderEditor.cs

```diff
@@ -528,47 +535,69 @@ void SetIsUtilityWindow(bool isUtilityWindow)
             res.titleContent = windowTitle;
         }
 
+#if UNITY_2020_2_OR_NEWER
+        static void UpdateToolContext()
+        {
+            if(!selectMode.IsMeshElementMode() && ToolManager.activeContextType != typeof(GameObjectToolContext))
+                ToolManager.SetActiveContext<GameObjectToolContext>();
+            if(selectMode.IsPositionMode() && ToolManager.activeContextType != typeof(PositionToolContext))
+                ToolManager.SetActiveContext<PositionToolContext>();
+            else if(selectMode.IsTextureMode() && ToolManager.activeContextType != typeof(TextureToolContext))
+                ToolManager.SetActiveContext<TextureToolContext>();
+        }
+#else
         internal static VertexManipulationTool activeTool
         {
             get
             {
                 return s_Instance == null
                     ? null
-                    : s_Instance.GetToolForSelectMode(s_Instance.m_CurrentTool, s_SelectMode);
+                    : (VertexManipulationTool)EditorToolManager.activeTool;
             }
         }
 
-        VertexManipulationTool GetTool<T>() where T : VertexManipulationTool, new()
+        void SetTool<T>() where T : VertexManipulationTool, new()
         {
-            VertexManipulationTool tool;
+            //If the type is already active do nothing
+            if(typeof(T) == ToolManager.activeToolType)
+                return;
+
+            VertexManipulationTool formerTool = null;
 
-            if (s_EditorTools.TryGetValue(typeof(T), out tool))
-                return tool;
-            tool = new T();
-            s_EditorTools.Add(typeof(T), tool);
-            return tool;
+            if(EditorToolManager.activeTool is VertexManipulationTool)
+                formerTool = (VertexManipulationTool)EditorToolManager.activeTool;
+            ToolManager.SetActiveTool(CreateInstance<T>());
+            if(formerTool != null)
+                DestroyImmediate(formerTool);
```

Comment:
`ToolManager.SetActiveTool<T>();`

`ToolManager.RestorePreviousPersistentTool`

Reply:
So using the Tool as a Singleton (with ToolManager.SetActiveTool<T>();) and not destroying it after use let the tool appear in the tool history.
![image](https://user-images.githubusercontent.com/66317117/91417108-96029200-e850-11ea-90b1-920acad20d7c.png)

However, tools from the history cannot be selected has the main tools are the ones that are considered.
So, what is the best behaviour to get on this ?

Reply:
This seems like something we should fix on trunk in the `EditorToolManager`

Reply:
Is it something to fix? 
As the tool is a singleton and not destroy, it's normal that it appears in there right ?

Reply:
That list is curated, and should only be showing tools that have an `EditorTool` attribute. 

I think it would also make sense to cull tools that resolve to BuiltIn tools, but I could see arguments to the contrary there.

Reply:
Agreed with the `EditorTool` attribute